### PR TITLE
Fix text alignment on block caret overlay

### LINKED ISSUE #2615: Block caret overlay misaligned with underlying text
The block caret actually draws the character underneath the caret onto an opaque text adornment and so any misalignment between the text rendering of the text view and the text rendering of the block caret is manifested as a slight jiggle in the character's position as the caret rolls over it or as the caret blinks.

Zooming the document window to 400% makes this easier to see:

image

Previously this alignment was nearly perfect but as of release 16.1.2 or so of Visual Studio 2019 (and a similar recent release of Visual Studio 2017), the jiggle is back.

While we are on the subject of the block caret, the block caret rendering does not work correctly when the caret is positioned over a character that is rendered as vim-style control character, e.g. `^C` for ASCII 3. In that case, the caret appears mostly solid depending on the text colors chosen for the caret and for control characters.

image

### PR DESCRIPTION
**Pull request by ricksladkey**
#### Issues

- Closes #2615

### CODE REVIEW

**Comment by ricksladkey** [comment]
Here is what the new caret alignment looks like:

image

and here is what the new control character display looks like:

image

**Comment by ricksladkey** [comment]
I don't edit files with tabs in them very often, but when I do, I wish the block caret floated over the last space of the tab like it does in vim. So I've rolled that feature into this PR:

image

File: Src/VimWpf/Implementation/BlockCaret/BlockCaret.cs [hunk 1]
```diff
@@ -338,36 +376,45 @@ private FormattedText CreateFormattedText()
...
-                        && !_controlCharUtil.IsDisplayControlChar(pointCharacter))
+                    if (_controlCharUtil.TryGetDisplayText(pointCharacter, out string caretString))
+                    {
+                        // Handle control character notation.
+                        caretCharacter = caretString;
+                        width = line.GetCharacterBounds(point).Width;
+                    }
+                    else if (Char.IsHighSurrogate(pointCharacter) &&
+                        point.Position < _textView.TextSnapshot.Length - 1 &&
+                        Char.IsLowSurrogate(point.Add(1).GetChar()))
```
Comment (jaredpar) [inline Src/VimWpf/Implementation/BlockCaret/BlockCaret.cs]: Consider using `point.Snapshot` vs `_textVIew.Snapshot`. Makes the code easier to read (don't have to verify they're the same snapshot.

File: Src/VimWpf/Implementation/BlockCaret/BlockCaret.cs [hunk 2]
```diff
@@ -338,36 +376,45 @@ private FormattedText CreateFormattedText()
...
                     var pointCharacter = point.GetChar();
-                    if (pointCharacter != '\t'
-                        && !_controlCharUtil.IsDisplayControlChar(pointCharacter))
+                    if (_controlCharUtil.TryGetDisplayText(pointCharacter, out string caretString))
+                    {
+                        // Handle control character notation.
+                        caretCharacter = caretString;
+                        width = line.GetCharacterBounds(point).Width;
+                    }
+                    else if (Char.IsHighSurrogate(pointCharacter) &&
```
Comment (jaredpar) [inline Src/VimWpf/Implementation/BlockCaret/BlockCaret.cs]: Think we should add a helper to `SnapshotPoinUtil` for this. Pretty sure we've written this code in a few places.

File: Src/VimWpf/Implementation/BlockCaret/BlockCaret.cs (hunk 1 above)
Comment (ricksladkey) [reply Src/VimWpf/Implementation/BlockCaret/BlockCaret.cs]: Excellent idea.

File: Src/VimWpf/Implementation/BlockCaret/BlockCaret.cs (hunk 2 above)
Comment (ricksladkey) [reply Src/VimWpf/Implementation/BlockCaret/BlockCaret.cs]: Converted it to use `SnapshotCodePoint.CodePointInfo`.

**Pull request merged**